README.md
NumChooseView
========

<img src="img\img_1.png" width="200" height="100"/>

<img src="img\img_3.gif" width="800" height="450"/>

Add NumChooseView to your project
----------------------------
Gradle:
```
   compile 'cn.yzapp.numchooseview:numchooseviewlib:[look download]@aar'
```

Maven:
```
<dependency>
  <groupId>cn.yzapp.numchooseview</groupId>
  <artifactId>numchooseviewlib</artifactId>
  <version>[look download]</version>
  <type>aar</type>
</dependency>
```
[ ![Download](https://api.bintray.com/packages/nesror/maven/NumChooseView/images/download.svg) ](https://bintray.com/artifact/download/nesror/maven/cn/yzapp/numchooseview/numchooseviewlib/1.5.5/numchooseviewlib-1.5.5.aar)

Use
----------------------------
 * 布局文件
````
<cn.yzapp.numchooseviewlib.NumChooseView
            android:id="@+id/num_view"
            android:layout_width="wrap_content"
            android:layout_height="30dp" />
````

 * 得到数量
~~~~
getBuyNum()
~~~~
 * 设置限制条件
~~~~
/**
  * 设置限制条件
  *
  * @param showStorage 库存
  * @param leastbuyNum 最少购买数量
  * @param limitNum    限购数量
  * @param basicNum    购买基数
  * NOT_LIMIT 不做限制
*/
setTerm(long showStorage, long leastbuyNum, long limitNum, long basicNum)
~~~~
 * 设置根视图，用于监听键盘是否收起
~~~~
setRootView(final View rootView)
~~~~
 * 设置是否显示提示
~~~~
setCanShowHint(boolean showToast)
~~~~

Release History
------------------------
[CHANGELOG](CHANGELOG.md)


CHANGELOG.md
# 1.5.0
* 增加购买基数
* 设置限制参数方式修改
~~~~
setShowStorage(long showStorage)
setLeastBuyNum(long buyNum)
setLimitNum(long limitNum)
~~~~
改为
~~~~
/**
  * 设置限制条件
  *
  * @param showStorage 库存
  * @param leastbuyNum 最少购买数量
  * @param limitNum    限购数量
  * @param basicNum    购买基数
*/
setTerm(long showStorage, long leastbuyNum, long limitNum, long basicNum)
~~~~
* demo修改


app/src/test/java/cn/yzapp/numchooseview/ExampleUnitTest.java
package cn.yzapp.numchooseview;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * To work on unit tests, switch the Test Artifact in the Build Variants view.
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

numchooseviewlib/src/test/java/cn/yzapp/numchooseviewlib/ExampleUnitTest.java
package cn.yzapp.numchooseviewlib;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * To work on unit tests, switch the Test Artifact in the Build Variants view.
 */
public class ExampleUnitTest {
    @Test
    public void addition_isCorrect() throws Exception {
        assertEquals(4, 2 + 2);
    }
}

app/src/main/java/cn/yzapp/numchooseview/MainActivity.java
package cn.yzapp.numchooseview;

import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.text.TextUtils;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;

import cn.yzapp.numchooseviewlib.NumChooseView;


public class MainActivity extends AppCompatActivity implements View.OnClickListener {

    private NumChooseView numChooseView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);
        numChooseView = (NumChooseView) findViewById(R.id.num_view);

        numChooseView.setRootView(findViewById(R.id.rl_root));
        numChooseView.setCanShowHint(true);

        findViewById(R.id.ok).setOnClickListener(this);
        findViewById(R.id.show_num).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                ((Button) view).setText("数量" + numChooseView.getBuyNum());
            }
        });

    }

    @Override
    public void onClick(View view) {
        numChooseView.setTerm(getNum(R.id.et_ShowStorage), getNum(R.id.et_LeastBuyNum), getNum(R.id.et_LimitNum), getNum(R.id.et_baseNum));
    }

    private long getNum(int res) {
        String num = ((EditText) findViewById(res)).getText().toString();
        if (TextUtils.isEmpty(num)) {
            return numChooseView.NOT_LIMIT;
        } else {
            return Long.parseLong(num);
        }

    }
}


numchooseviewlib/src/main/java/cn/yzapp/numchooseviewlib/model/NumBean.java
package cn.yzapp.numchooseviewlib.model;

import cn.yzapp.numchooseviewlib.NumChooseView;

/**
 * @author: nestor
 * time: 1/22 022-13:58.
 * email: nestor@yzapp.cn
 * desc:
 */
public class NumBean {
    private long showStorage = NumChooseView.NOT_LIMIT;//库存
    private long leastBuyNum = 1;//最少购买数量
    private long limitNum = NumChooseView.NOT_LIMIT;//限购数量
    private long basicNum = 1;//购买基数

    public long getShowStorage() {
        return showStorage;
    }

    public void setShowStorage(long showStorage) {
        this.showStorage = showStorage;
    }

    public long getLeastBuyNum() {
        return leastBuyNum;
    }

    public void setLeastBuyNum(long leastBuyNum) {
        this.leastBuyNum = leastBuyNum;
    }

    public long getLimitNum() {
        return limitNum;
    }

    public void setLimitNum(long limitNum) {
        this.limitNum = limitNum;
    }

    public long getBasicNum() {
        return basicNum;
    }

    public void setBasicNum(long basicNum) {
        this.basicNum = basicNum;
    }
}


numchooseviewlib/src/main/java/cn/yzapp/numchooseviewlib/NumTextWatcher.java
package cn.yzapp.numchooseviewlib;

import android.annotation.SuppressLint;
import android.content.Context;
import android.graphics.Color;
import android.text.Editable;
import android.text.TextWatcher;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.Toast;

import cn.yzapp.numchooseviewlib.model.NumBean;

/**
 * @author: nestor
 * time: 1/22 022-13:58.
 * email: nestor@yzapp.cn
 * desc:
 */
public class NumTextWatcher implements TextWatcher {
    private final EditText tvNum;
    private final ImageView numLes;
    private final ImageView numAdd;
    private final Context mContext;
    private NumBean mNumBean;
    private long mGoodNum;

    public NumTextWatcher(Context context, NumBean numBean, long goodNum, EditText tvNum, ImageView numLes, ImageView numAdd) {
        this.mContext = context;
        this.mNumBean = numBean;
        this.mGoodNum = goodNum;
        this.tvNum = tvNum;
        this.numLes = numLes;
        this.numAdd = numAdd;
    }

    @Override
    public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @Override
    public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

    }

    @SuppressLint("SetTextI18n")
    @Override
    public void afterTextChanged(Editable s) {
        if (mNumBean.getShowStorage() < mNumBean.getLeastBuyNum() || mNumBean.getLimitNum() < mNumBean.getLeastBuyNum()) {
            return;
        }
        //删除数量－置灰
        if (s.length() < 1) {
            numLes.setBackgroundResource(R.color.bg_gray);
            mGoodNum = -1;
            tvNum.setSelection(tvNum.length());
            return;
        }

        if (mNumBean != null) {
            if (mNumBean.getLimitNum() != -1 && Long.parseLong(s.toString()) > mNumBean.getLimitNum()) {
                numAdd.setBackgroundResource(R.color.bg_gray);
                tvNum.setText(mNumBean.getLimitNum() + "");
                NumChooseView.showToast(mContext, "不能超过限购数量！");
                tvNum.setSelection(tvNum.length());
                mGoodNum = mNumBean.getLimitNum();
                return;
            }

            if (mNumBean.getShowStorage() != -1 && Long.parseLong(s.toString()) > mNumBean.getShowStorage()) {
                numAdd.setBackgroundResource(R.color.bg_gray);
                tvNum.setText(mNumBean.getShowStorage() + "");
                NumChooseView.showToast(mContext, "不能超过库存数量！");
                tvNum.setSelection(tvNum.length());
                mGoodNum = mNumBean.getShowStorage();
                return;
            }
        }
        mGoodNum = Long.parseLong(s.toString());
        if (mGoodNum <= 1 || (mNumBean.getLeastBuyNum() != -1 && mGoodNum <= mNumBean.getLeastBuyNum())) {
            numLes.setBackgroundResource(R.color.bg_gray);
        } else {
            numLes.setBackgroundColor(Color.TRANSPARENT);
        }

        if ((mNumBean.getLimitNum() != -1 && mGoodNum >= mNumBean.getLimitNum()) || (mNumBean.getShowStorage() != -1 && mGoodNum >= mNumBean.getShowStorage())) {
            numAdd.setBackgroundResource(R.color.bg_gray);
        } else {
            numAdd.setBackgroundColor(Color.TRANSPARENT);
        }
    }
}


numchooseviewlib/src/main/java/cn/yzapp/numchooseviewlib/NumChooseView.java
package cn.yzapp.numchooseviewlib;

import android.annotation.SuppressLint;
import android.annotation.TargetApi;
import android.content.Context;
import android.graphics.Color;
import android.os.Build;
import android.text.Editable;
import android.text.TextUtils;
import android.text.TextWatcher;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewTreeObserver;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.LinearLayout;

import cn.yzapp.numchooseviewlib.model.NumBean;
import cn.yzapp.numchooseviewlib.util.ToastUtil;

/**
 * @author: nestor
 * time: 1/22 022-13:58.
 * email: nestor@yzapp.cn
 * desc:
 */
public class NumChooseView extends LinearLayout implements View.OnClickListener {
    public final static int NOT_LIMIT = -1;
    private NumBean mNumBean;
    private long mGoodNum = 1;
    private ImageView numAdd;
    private EditText tvNum;
    private ImageView numLes;
    private boolean showToast;

    public NumChooseView(Context context) {
        super(context);
        initView();
    }

    public NumChooseView(Context context, AttributeSet attrs) {
        super(context, attrs);
        initView();
    }

    public NumChooseView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        initView();
    }

    @TargetApi(Build.VERSION_CODES.LOLLIPOP)
    public NumChooseView(Context context, AttributeSet attrs, int defStyleAttr, int defStyleRes) {
        super(context, attrs, defStyleAttr, defStyleRes);
        initView();
    }

    /**
     * 设置限制条件
     *
     * @param showStorage 库存
     * @param leastbuyNum 最少购买数量
     * @param limitNum    限购数量
     * @param basicNum    购买基数
     */
    public void setTerm(long showStorage, long leastbuyNum, long limitNum, long basicNum) {
        setShowStorage(showStorage);
        setLeastBuyNum(leastbuyNum);
        setLimitNum(limitNum);
        setBasicNum(basicNum);
        check();
    }

    boolean rootViewCheck;

    /**
     * 设置根视图，用于监听键盘是否收起
     *
     * @param rootView 最少购买数量
     */
    public void setRootView(final View rootView) {
        setRootViewCheck(rootView);
    }

    /**
     * 设置是否显示提示
     */
    public void setCanShowHint(boolean showToast) {
        this.showToast = showToast;
    }

    /**
     * 得到数量
     */
    public long getBuyNum() {
        return mGoodNum;
    }

    /**
     * 得到显示数量
     */
    public long getShowNum() {
        if (TextUtils.isEmpty(tvNum.getText().toString())) {
            return 0;
        }
        return Long.parseLong(tvNum.getText().toString());
    }

    private void setShowStorage(long showStorage) {
        mNumBean.setShowStorage(showStorage);
    }

    private void setLeastBuyNum(long buyNum) {
        if (buyNum < 1) {
            buyNum = 1;
        }
        mNumBean.setLeastBuyNum(buyNum);
        mGoodNum = buyNum;
        tvNum.setText("" + mGoodNum);
    }

    private void setLimitNum(long limitNum) {
        mNumBean.setLimitNum(limitNum);
    }

    private void setBasicNum(long basicNum) {
        if (basicNum < 1) {
            basicNum = 1;
        }
        mNumBean.setBasicNum(basicNum);
    }

    private void initView() {
        mNumBean = new NumBean();
        View numView = View.inflate(getContext(), R.layout.numchoose_include_view_num_choose, null);
        tvNum = (EditText) numView.findViewById(R.id.tv_num);
        numLes = (ImageView) numView.findViewById(R.id.num_les);
        numAdd = (ImageView) numView.findViewById(R.id.num_add);
        tvNum.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {

            }

            @Override
            public void afterTextChanged(Editable editable) {
                //删除数量－置灰
                if (editable.length() < 1) {
                    numLes.setBackgroundResource(R.color.numchoose_bg_gray);
                    tvNum.setSelection(tvNum.length());
                    return;
                }

                if (mNumBean != null) {
                    if (mNumBean.getLimitNum() != NOT_LIMIT && Long.parseLong(editable.toString()) > mNumBean.getLimitNum()) {
                        numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
                        tvNum.setText((mGoodNum < mNumBean.getLimitNum() ? mGoodNum : mNumBean.getLimitNum()) + "");
                        showToast(getContext(), "不能超过限购数量！");
                        tvNum.setSelection(tvNum.length());
                        return;
                    }

                    if (mNumBean.getShowStorage() != NOT_LIMIT && Long.parseLong(editable.toString()) > mNumBean.getShowStorage()) {
                        numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
                        tvNum.setText((mGoodNum < mNumBean.getShowStorage() ? mGoodNum : mNumBean.getShowStorage()) + "");
                        showToast(getContext(), "不能超过库存数量！");
                        tvNum.setSelection(tvNum.length());
                        return;
                    }
                }
                if (mNumBean.getBasicNum() == 1 || (Long.parseLong(editable.toString()) % mNumBean.getBasicNum()) == 0) {
                    if (Long.parseLong(editable.toString()) >= mNumBean.getLeastBuyNum()) {
                        mGoodNum = Long.parseLong(editable.toString());
                    }
                }
                if (mGoodNum <= 1 || (mNumBean.getLeastBuyNum() != NOT_LIMIT && mGoodNum <= mNumBean.getLeastBuyNum())) {
                    numLes.setBackgroundResource(R.color.numchoose_bg_gray);
                } else {
                    numLes.setBackgroundColor(Color.TRANSPARENT);
                }

                if ((mNumBean.getLimitNum() != NOT_LIMIT && mGoodNum >= mNumBean.getLimitNum()) || (mNumBean.getShowStorage() != NOT_LIMIT && mGoodNum >= mNumBean.getShowStorage())) {
                    numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
                } else {
                    numAdd.setBackgroundColor(Color.TRANSPARENT);
                }
            }
        });
        numLes.setOnClickListener(this);
        numAdd.setOnClickListener(this);

        addView(numView);

    }

    @SuppressLint("SetTextI18n")
    @Override
    public void onClick(View view) {
        if (view.getId() == R.id.num_les) {
            if (mNumBean.getLeastBuyNum() == NOT_LIMIT || (mGoodNum - mNumBean.getBasicNum()) >= (mNumBean.getLeastBuyNum() == 0 ? 1 : mNumBean.getLeastBuyNum())) {
                mGoodNum -= mNumBean.getBasicNum();
                tvNum.setText("" + mGoodNum);
            } else {
                numLes.setBackgroundResource(R.color.numchoose_bg_gray);
                showToast(getContext(), "不能少于起购数量！");
            }
        } else if (view.getId() == R.id.num_add) {
            if (mNumBean.getLimitNum() == NOT_LIMIT || ((mGoodNum + mNumBean.getBasicNum()) <=
                    mNumBean.getLimitNum())) {
                if (mNumBean.getShowStorage() == NOT_LIMIT || mGoodNum < mNumBean.getShowStorage()) {
                    mGoodNum += mNumBean.getBasicNum();
                    tvNum.setText("" + mGoodNum);
                    numLes.setBackgroundColor(Color.TRANSPARENT);
                } else {
                    numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
                    showToast(getContext(), "不能超过库存数量！");
                }
            } else {
                numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
                showToast(getContext(), "不能超过限购数量！");
            }
        }
    }

    private void setRootViewCheck(final View rootView) {
        rootView.getViewTreeObserver().addOnGlobalLayoutListener(
                new ViewTreeObserver.OnGlobalLayoutListener() {
                    @Override
                    public void onGlobalLayout() {
                        int heightDiff = rootView.getRootView().getHeight() - rootView.getHeight();

                        if (!rootViewCheck && heightDiff < 300) { // 键盘收起
                            rootViewCheck = true;
                            if (TextUtils.isEmpty(tvNum.getText().toString())) {
                                tvNum.setText(mGoodNum + "");
                                tvNum.setSelection(tvNum.length());
                                return;
                            }
                            if (mNumBean.getBasicNum() != 1 && (Long.parseLong(tvNum.getText().toString()) % mNumBean.getBasicNum()) != 0) {
                                showToast(getContext(), "必须是购买基数的整数倍！");
                                tvNum.setText(mGoodNum + "");
                                tvNum.setSelection(tvNum.length());
                            } else if ((mNumBean.getLeastBuyNum() != -1 && Long.parseLong(tvNum.getText().toString()) < mNumBean.getLeastBuyNum())) {
                                //不能小于起购数量－置灰
                                numLes.setBackgroundResource(R.color.numchoose_bg_gray);
                                showToast(getContext(), "不能少于起购数量！");
                                tvNum.setText(mGoodNum + "");
                                /*if (mNumBean.getLeastBuyNum() != 0) {
                                    tvNum.setText(mNumBean.getLeastBuyNum() + "");
                                    mGoodNum = mNumBean.getLeastBuyNum();
                                } else {
                                    tvNum.setText("1");
                                    mGoodNum = 1;
                                }*/
                                tvNum.setSelection(tvNum.length());
                            }
                        } else {
                            rootViewCheck = false;
                        }
                    }
                });
    }

    private void showToast(Context ct, String s) {
        if (showToast) {
            ToastUtil.shortToast(ct, s);
        }
    }

    private void check() {
        if (mNumBean.getLimitNum() != NOT_LIMIT && (mNumBean.getLimitNum() < mNumBean.getLeastBuyNum())) {
            showToast(getContext(), "最大购买数量不能小于最低购买数量！");
            tvNum.setText(mNumBean.getLimitNum() + "");
            mGoodNum = mNumBean.getLimitNum();
            numLes.setBackgroundResource(R.color.numchoose_bg_gray);
            numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
            numLes.setEnabled(false);
            numAdd.setEnabled(false);
            tvNum.setEnabled(false);
            return;
        }

        if (mNumBean.getBasicNum() > mNumBean.getLeastBuyNum()) {
            numLes.setBackgroundResource(R.color.numchoose_bg_gray);
            tvNum.setText(mNumBean.getBasicNum() + "");
            mGoodNum = mNumBean.getBasicNum();
        } else if (mNumBean.getLeastBuyNum() % mNumBean.getBasicNum() != 0) {
            long num = mNumBean.getBasicNum() * (mNumBean.getLeastBuyNum() / mNumBean.getBasicNum() + 1);

            if (mNumBean.getLimitNum() != NOT_LIMIT && (mNumBean.getLimitNum() < num)) {
                tvNum.setText(mNumBean.getLimitNum() + "");
                mGoodNum = mNumBean.getLimitNum();
                numLes.setBackgroundResource(R.color.numchoose_bg_gray);
                numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
                numLes.setEnabled(false);
                numAdd.setEnabled(false);
                tvNum.setEnabled(false);
                return;
            }
            if (mNumBean.getShowStorage() != NOT_LIMIT && (mNumBean.getShowStorage() < num)) {
                tvNum.setText(mNumBean.getShowStorage() + "");
                mGoodNum = mNumBean.getShowStorage();
                numLes.setBackgroundResource(R.color.numchoose_bg_gray);
                numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
                numLes.setEnabled(false);
                numAdd.setEnabled(false);
                tvNum.setEnabled(false);
            } else {
                numLes.setBackgroundResource(R.color.numchoose_bg_gray);
                tvNum.setText(num + "");
                mGoodNum = num;
            }

        }

        if (mNumBean.getShowStorage() != NOT_LIMIT && (mNumBean.getShowStorage() < mNumBean.getLeastBuyNum() || mNumBean.getShowStorage() < mNumBean.getBasicNum())) {
            tvNum.setText(mNumBean.getShowStorage() + "");
            mGoodNum = mNumBean.getShowStorage();
            numLes.setBackgroundResource(R.color.numchoose_bg_gray);
            numAdd.setBackgroundResource(R.color.numchoose_bg_gray);
            numLes.setEnabled(false);
            numAdd.setEnabled(false);
            tvNum.setEnabled(false);
        }

    }

}
